### DISCUSSION TITLE
Marker bei WMS Diensten setzen

### DISCUSSION THREAD

**Discussion by nicole-vlbg** [e-netze/webgis-community#77] [Q&A]
Wenn man in einer Karte mit einem WMS Service (in diesem Fall ein GeoServer Service) ein Objekt abfragt (identify) werden die Attribute dargestellt und neben der Tabelle ein Marker (siehe Screenshot).
Allerdings wird an der Stelle des Klicks kein Marker gesetzt.
Man weiß nicht mehr, wo man geklickt hat.

Gibt es die Möglichkeit an dieser Stelle einen Marker zu setzen?

![bsp_abfrage_wms (002)](https://github.com/gis-eni/webgis-forum/assets/124671123/a3aa8326-7eb9-4795-8d4d-8a31ea2319ce)

**Comment by gis-eni** [comment]
Bei Flächen ist das theoretisch möglich. Allerdings kann das bei Punkten oder Linien zu falsch-Interpretationen kommen. Hier muss der Klickpunkt nicht zwingend auf der Linine/Punkt liegen, sondern ist nur im aktuellen Maßstabs innerhalb der Klicktoleranz beim Objekt. Wenn der Anwender danach in die Karte weiter hinein zoomt, kann der Marker relativ weit vom eigentlichen Objekte entfernt liegen => wird eher verwirrend.

Workaround: Der Dienst sollte auch die Geometrie zurück liefern. 
Die meisten WMS Dienst können auch einen FeatureResultType "application/json" oder "geojson" zurück geben. Das kann über die Capabilitities überprüft werden. Da wird dann in der Regel auch die Objekt Geometrie an den Client übergeben. Dann wird auch in der Karte ein Marker "an einer gültigen" Stelle angezeigt.

PS: Die Geometrie wird bei einen der folgenden **GetFeatureInfoFormaten" ausgelesen:

```
application/json
application/geojson
application/geo+json
```

Hier immer den Typ verwenden, der vom Dienst bei den Capabilities angeben wird.
Sollte bei diesem Type keine Geometrie zurück kommen, setzt WebGIS trotzdem einen Marker an den Klickpunkt. Das ist praktisch, wenn der Layer einen Rasterthema ist, dass keine Geometrie zurück geben würde. Hier gibt es keine "Toleranz", da diese Themen flächenfüllen sind.

Bei allen anderen XML/Text Formaten wird aus oben genannten Gründen kein Marker gesetzt.

Wenn man ein Json-Format als InfoFormat angibt, muss man sich im CMS auch um die Tabelle kümmern. Spalten müssen einzeln angeben werden. Ein Workaround ist hier, dass man für die Tabelle im CMS eine Spalte (zB **All**) anlegt. Als ``Feld Name`` kann für dieses Spalte ``*`` angeben. Damit werden alle Werte aus dem JSON angezeigt (das geht grundsätzlich für alle Abfragen, nicht nur bei WMS Diensten)

**Reply by faulgurke** [reply]
Hmm, wir haben uns an die Anleitung gehalten, aber bekommen die Meldung: **field ignored (not found in layer)**

Es gibt also im GeoServer Verzeichnis des Layers die drei Dateien header_json.ftl, content_json.ftl, footer_json.ftl.
CMS Tabellendefinition: Spalten Typ "Field", Name "*"

getCapabilities:
<GetFeatureInfo>
<Format>text/plain</Format>
<Format>application/vnd.ogc.gml</Format>
<Format>text/xml</Format>
<Format>application/vnd.ogc.gml/3.1.1</Format>
<Format>text/xml; subtype=gml/3.1.1</Format>
<Format>text/html</Format>
**<Format>application/json</Format>**

CMS Service Einstellungen -> WMS GetFeatureInfo Format:
application/json

und die Anleitung auf der GeoServer Doku befolgt:
https://docs.geoserver.org/latest/en/user/tutorials/GetFeatureInfo/geojson.html

Fällt dir irgendetwas auf?

**Comment by gis-eni** [comment]
Kann ich wenig dazu sagen. Bräuchte ich zugriff auf den Dient und mehr Details: Was möchtest du abfragen, usw.
Dann auch per Mail sein, wenn du den Link zum Dienst nicht hier posten willst...
